File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int size = s.length(), left = 0, maxLen = 0;
        set<int> chars;
        for(int right = 0 ; right < size ; right++) {

            while(chars.contains(s[right])) {
                chars.erase(s[left]);
                left++;
            }

            chars.insert(s[right]);
            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int maxLen = 0, len = s.length(), left = 0;

        for(int right = 0; right < len; right++) {
            char ch = s.charAt(right);

            if(map.containsKey(ch))
                left = Math.max(left, map.get(ch) + 1);

            map.put(ch, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}